File: src/chapters/chapter-05/csharp/NativePlugins/UserProfilePlugin.cs

using Microsoft.SemanticKernel;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.DependencyInjection;

using System.ComponentModel;
using System.Collections.Generic;
using System.Text.Json.Serialization;

using AdvancedAIShoppingAssistant.Models;
using AdvancedAIShoppingAssistant.Services;

namespace AdvancedAIShoppingAssistant.NativePlugins;

public class UserProfilePlugin
{
    private readonly ILogger<UserProfilePlugin> _logger;
    private readonly UserProfileService _userProfileService;

    /// <summary>
    /// Initializes a new instance of the UserProfilePlugin class with logging and user profile service.
    /// </summary>
    /// <param name="logger">The logger instance.</param>
    /// <param name="userProfileService">The user profile service.</param>
    public UserProfilePlugin(ILogger<UserProfilePlugin> logger, UserProfileService userProfileService)
    {
        _logger = logger;
        _userProfileService = userProfileService;
    }

    /// <summary>
    /// Retrieves the budget limit for a specific user.
    /// </summary>
    /// <param name="username">The username of the user.</param>
    /// <returns>The budget preference or an error message.</returns>
    [KernelFunction("get_budget_limit")]
    [Description("Retrieves the budget limit for a specific user.")]
    public string GetBudgetLimit([Description("The user id of the user.")] string userId)
    {
        if (!_userProfileService.UserProfiles.ContainsKey(userId))
        {
            return $"User with username '{userId}' does not exist.";
        }

        var profile = _userProfileService.UserProfiles[userId];

        return $"Budget for {userId}: $ {profile.Budget}";
    }

    /// <summary>
    /// Retrieves the brand affinity for a specific user.
    /// </summary>
    /// <param name="userId">The user id of the user.</param>
    /// <returns>The brand affinity or an error message.</returns>
    [KernelFunction("get_brand_affinity")]
    [Description("Retrieves the brand affinity for a specific user.")]
    public string GetBrandAffinity(
        [Description("The user Id of the user.")] string userId)
    {
        if (!_userProfileService.UserProfiles.ContainsKey(userId))
        {
            return $"User with username '{userId}' does not exist.";
        }

        var profile = _userProfileService.UserProfiles[userId];

        return $"Brand Affinity for {userId}: {profile.BrandAffinity}";
    }

    /// <summary>
    /// Retrieves the category interests for a specific user.
    /// </summary>
    /// <param name="userId">The user id of the user.</param>
    /// <returns>The category interests or an error message.</returns>
    [KernelFunction("get_category_interests")]
    [Description("Retrieves the categories of interests for a specific user.")]
    public string GetCategoryInterests(
        [Description("The user id of the user.")] string userId)
    {
        if (!_userProfileService.UserProfiles.ContainsKey(userId))
        {
            return $"User with username '{userId}' does not exist.";
        }

        var profile = _userProfileService.UserProfiles[userId];
        return $"Category Interests for {userId}: {string.Join(", ", profile.CategoryInterests)}";
    }

    /// <summary>
    /// Retrieves the email address for a specific user.
    /// </summary>
    /// <param name="userId">The user id of the user.</param>
    /// <returns>The email address of the user. </returns>
    [KernelFunction("get_email_address")]
    [Description("Retrieves the email address for a specific user.")]
    public string GetEmailAddress([Description("The user id of the user.")] string userId)
    {
        if (!_userProfileService.UserProfiles.ContainsKey(userId))
        {
            return $"User with username '{userId}' does not exist.";
        }

        var profile = _userProfileService.UserProfiles[userId];
        return $"Email Address for {userId}: {profile.Email}";
    }

    /// <summary>
    /// Retrieves the latest 5 visited products for a specific user.
    /// </summary>
    /// <param name="userId">The user id of the user.</param>
    /// <returns>The latest 5 visited products or an error message.</returns>
    [KernelFunction("get_latest_visited_products")]
    [Description("Retrieves the latest visited products for a specific user.")]
    public string GetLatestVisitedProducts(
        [Description("The user id of the user.")] string userId)
    {
        if (!_userProfileService.UserProfiles.ContainsKey(userId))
        {
            return $"User with username '{userId}' does not exist.";
        }

        var profile = _userProfileService.UserProfiles[userId];

        return $"Latest Visited Products for {userId}: {string.Join(", ", profile.LatestVisitedProducts)}";
    }
}

File: src/chapters/chapter-05/python/plugins/user_profile_plugin.py
import logging

from services.user_profile_service import UserProfileService

from semantic_kernel.functions import kernel_function
from typing import List, Optional
from dataclasses import dataclass, field

class UserProfilePlugin:
    def __init__(self, user_profile_service: UserProfileService, logger: logging.Logger):
        self._logger = logger
        self._user_profile_service = user_profile_service

    # #A Method to retrieve the budget limit for a user
    # Retrieves the budget limit for a specific user.
    # Arguments:
    # - user_id: The ID of the user whose budget limit is to be retrieved.
    @kernel_function(name="get_budget_limit", description="Retrieves the budget limit for a specific user.")
    def get_budget_limit(self, user_id: str) -> str:
        if user_id not in self._user_profile_service.user_profiles:
            return f"User with username '{user_id}' does not exist."

        profile = self._user_profile_service.user_profiles[user_id]
        return f"Budget for {user_id}: $ {profile.budget}"

    # #B Method to retrieve the brand affinity for a user
    # Retrieves the brand affinity for a specific user.
    # Arguments:
    # - user_id: The ID of the user whose brand affinity is to be retrieved.
    @kernel_function(name="get_brand_affinity", description="Retrieves the brand affinity for a specific user.")
    def get_brand_affinity(self, user_id: str) -> str:
        if user_id not in self._user_profile_service.user_profiles:
            return f"User with username '{user_id}' does not exist."

        profile = self._user_profile_service.user_profiles[user_id]
        return f"Brand Affinity for {user_id}: {profile.brand_affinity}"

    # #C Method to retrieve the category interests for a user
    # Retrieves the category interests for a specific user.
    # Arguments:
    # - user_id: The ID of the user whose category interests are to be retrieved.
    @kernel_function(name="get_category_interests", description="Retrieves the categories of interests for a specific user.")
    def get_category_interests(self, user_id: str) -> str:
        if user_id not in self._user_profile_service.user_profiles:
            return f"User with username '{user_id}' does not exist."

        profile = self._user_profile_service.user_profiles[user_id]
        return f"Category Interests for {user_id}: {', '.join(profile.category_interests)}"

    # #D Method to retrieve the email address for a user
    # Retrieves the email address for a specific user.
    # Arguments:
    # - user_id: The ID of the user whose email address is to be retrieved.
    @kernel_function(name="get_email_address", description="Retrieves the email address for a specific user.")
    def get_email_address(self, user_id: str) -> str:
        if user_id not in self._user_profile_service.user_profiles:
            return f"User with username '{user_id}' does not exist."

        profile = self._user_profile_service.user_profiles[user_id]
        return f"Email Address for {user_id}: {profile.email}"

    # #E Method to retrieve the latest visited products for a user
    # Retrieves the latest 5 visited products for a specific user.
    # Arguments:
    # - user_id: The ID of the user whose latest visited products are to be retrieved.
    @kernel_function(name="get_latest_visited_products", description="Retrieves the latest visited products for a specific user.")
    def get_latest_visited_products(self, user_id: str) -> str:
        if user_id not in self._user_profile_service.user_profiles:
            return f"User with username '{user_id}' does not exist."

        profile = self._user_profile_service.user_profiles[user_id]
        return f"Latest Visited Products for {user_id}: {', '.join(profile.latest_visited_products)}"

